File: Medium/Container_With_Most_Water.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Sep 10 20:59:27 2022

@author: jamescrisp
"""

class Solution:
    def maxArea(self, height: List[int]) -> int:
        
        #Maxmimum area calculation
        
        #How to calculate without iterating through each and ever volume?
        
        i = 0
        j = len(height)-1
        water = 0
        
        while i != j:
            water = max(water,(j-i)*min(height[i],height[j]))
            
            if height[i] <= height[j]:
                i+=1
            else:
                j-=1
        
        return water
            
            

File: Medium/Container_With_Most_Water.java
/**
 * Find max volume of water a container can hold
 * 
 * @author James Crisp
 */
class Solution {
    public int maxArea(int[] height) {
        
        // Index of the left side container
        int iLeft = 0;
        // Index of the right side of the container
        int iRight = 1;
        // Max volume the container can hold
        int maxVol = (iRight - iLeft) * min(height[iLeft], height[iRight]);

        for (int i = 2; i < height.length() - 1; i++) {
            if (maxVol < (i - iLeft) * min(height(iLeft), height(i))) {
                maxVol = (i - iLeft) * min(height(iLeft), height(i));
                iRight = i;
            }
            if (maxVol < (i - iRight) * min(height(iRight), height(i))) {
                maxVol = (i - iRight) * min(height(iRight), height(i));
                iLeft = iRight;
                iRight = i;
            }
        }

        return maxVol;
    }
}